osu.Game/Online/Multiplayer/MultiplayerCountdown.cs

```diff
@@ -0,0 +1,24 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable enable
+
+using System;
+using MessagePack;
+
+namespace osu.Game.Online.Multiplayer
+{
+    /// <summary>
+    /// Describes the current countdown in a <see cref="MultiplayerRoom"/>.
+    /// </summary>
+    [MessagePackObject]
+    [Union(0, typeof(MatchStartCountdown))] // IMPORTANT: Add rules to SignalRUnionWorkaroundResolver for new derived types.
+    public abstract class MultiplayerCountdown
+    {
+        /// <summary>
+        /// The time at which the countdown will end.
+        /// </summary>
+        [Key(0)]
+        public DateTimeOffset EndTime { get; set; }
```

Comment:
Before I start to review these PRs, on a quick check this looks to be relying completely on the fact that clocks are synced perfect between the client and server. Which is almost never going to be the case. I can't find any mention of this in the PR chain descriptions, potentially an oversight?

Reply:
The client only sends a duration to the server. The server has the sole responsibility for deciding the end time for the clients to update their displays with.

Unless I'm missing something (and I could be?), the client displays should be accurate and not relying on perfect syncing.

There's a little bit of delay at the point of starting a match, just because of server-to-client message latency + the time it takes to run the async callback, which hasn't been compensated for and is quite involved to do, but I don't think it's worth it. It's going to be <1sec delay from anywhere in the world with a decent (no packet loss) connection.

Reply:
https://github.com/ppy/osu/blob/4630aa15cc06a1efd1df0fe1486777c50d6999db/osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerReadyButton.cs#L289

This line is doing subtraction of local system time from server time. If the user's clock is out of sync, it may be slightly off. If the user's clock has the wrong time zone or if daylight savings has changed recently and not updated, it may be very off.

Reply:
Just to confirm. The issue then is that the server sends `DateTimeOffset.Now` and the client uses `DateTimeOffset.UtcNow`? I was under the impression that this would be fine since `DateTimeOffset` stores a `DateTime` plus an offset, so I can't see a reason why it would fail.

From the MSDN docs on `DateTimeOffset`:

> You can add or subtract either dates or time intervals from a particular [DateTimeOffset](https://docs.microsoft.com/en-us/dotnet/api/system.datetimeoffset?view=net-6.0) value. Arithmetic operations with [DateTimeOffset](https://docs.microsoft.com/en-us/dotnet/api/system.datetimeoffset?view=net-6.0) values, unlike those with [DateTime](https://docs.microsoft.com/en-us/dotnet/api/system.datetime?view=net-6.0) values, **adjust for differences in time offsets when returning a result**.

(Emphasis mine).

Would it appease you if I changed both to use `UtcNow`? Or what is your suggestion?

Reply:
It's nothing to do with UTC. It's the fact that the two clocks are not the same clock, and therefore *will* be out of sync by seconds or minutes.

If the user's clock is reading UTC12:21:05 and the server is UTC12:21:10 a countdown of 5 seconds is going to show as 10 seconds to the local user, causing a bit of confusion.

Reply:
Well, what's the solution there then? Send a message for every second from the server? Bancho already struggles with this, where it can skip messages during the final seconds of the match start countdown (it sends starting in 5..4..3.. etc), so I'd like to avoid that.

Furthermore, how do you deal with users joining rooms while a countdown is running?

Reply:
Not every second. The countdown object could have a start time inside the object (if the only case is transmitting a counter start event) or a current time (given here we potentially have users joining halfway through). Or even simpler, the value sent to the client could be `double SecondsRemaining`.

Reply:
I'd like to avoid sending a `SecondsRemaining` var. It means users joining rooms during a countdown will have the incorrect time, unless you send the message/update the room basically every second.

Sending a start `DateTimeOffset` sounds fine?

Reply:
> I'd like to avoid sending a `SecondsRemaining` var. It means users joining rooms during a countdown will have the incorrect time, unless you send the message/update the room basically every second.

Obviously you'd need to send the correct time at the point of sending the message to the client.

> Sending a start `DateTimeOffset` sounds fine?

Again, this will *only* work for the case where you have an explicit `CountdownStarted` event, and even then you will need to use the start and end time to create a new duration value. It won't solve the joining-midway case. Basically you're sending a `CountdownDuration` via a much more complicated data structure (`start` + `end` `DateTimeOffset`s).

Reply:
Alright, I've changed it to send a `TimeRemaining` instead. It's only sent upon `JoinRoom()`/`CountdownChangedEvent` because that's the only time when the room is serialised to the client and/or the only way clients receive countdown notifications.